nexus/src/external_api/http_entrypoints.rs

```diff
@@ -2014,6 +2015,37 @@ async fn instance_reboot(
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
 
+/// Provision an instance
+#[endpoint {
+    method = POST,
+    path = "/organizations/{organization_name}/projects/{project_name}/instances/{instance_name}/provision",
+    tags = ["instances"],
+}]
+async fn instance_provision(
+    rqctx: Arc<RequestContext<Arc<ServerContext>>>,
+    path_params: Path<InstancePathParam>,
```

Comment:
I don't think I understand what "provisioned" means in that I'm not sure why this doesn't require the same parameters as `instance_create` with regard to, say, memory or CPU count.

Reply:
I sorta agree, I do find `provisioned` a bit clunky here. it'd perhaps make more sense in a context where we might want to distinguish for billing purposes.

I'm thinking just removing the provisioned/provisioning states, and giving the start endpoint an optional query param (e.g. `/instances/foo/start?hold=true`) that would go through the same motions but stop short at telling the vCPUs to run. The user-visible state at the point would just be `Starting` until a subsequent `start` request.

The cli then could have an optional `--hold` flag for `instance start`. The wording on the console side still feels a bit tricky to get right. Maybe `Start (Held)`? Or is it just reasonable to say this is a rare enough thing already that leaving it cli only is ok?

@david-crespo @zephraph I know y'all mentioned not liking reusing `/start` with a body but does the above seem reasonable?

Reply:
> Or is it just reasonable to say this is a rare enough thing already that leaving it cli only is ok?

We _could_ suppress it from being present in Console UI, but one product principle we're trying to adhere to is that it should be seamless to move from one client to another. Have an option available in one place that's not available in another can happen but generally just introduces confusion and we'd like to avoid it if possible.

I guess the real question I have is this: Is provisioned a real state? If we're trying to express that there's a state being `creating` and `starting` maybe that's just `created`? Overall I think I agree with one of @ahl's concerns here which is that the more states we have the harder it is to educate users about what each state is and what they need to do. 

I think it would be easier to communicate that an instance has been created but not yet started then it would be to say you selected not to start an instance so now it's in this provisioned state that you normally don't ever see. However, if `provisioned` is a real state and you need to be able to stop and instance and get it back to the provisioned state then it is what it is. 

I still hold the position that the `/start` action should only ever be able to start the instance.

Reply:
So, here are the states as they exist in this PR:

- **Creating**: Initial state an instance starts in during instance creation saga and up until the last saga action when sled agent beings setting up the instance.
- **Provisioning**: Instance is being set up on a sled (creating propolis zone, opte ports, attaching, creating the instance in propolis, etc)
- **Provisioned**:  Instance is up and ready on a sled (disks attached, opte port created, propolis zone up and running, etc) but the vCPUs are not yet running.
- **Starting**: Propolis' `Boot` state.
- **Running**: Instance is running.
- **Stopping**: Instance is stopping (shutting down the guest, deattaching disks, destroying propolis zone, etc)
- **Stopped**: Instance exists in the DB but not currently running on any sled.
- **Rebooting**: Instance is rebooting.
- **Migrating**: Instance is being migrated.
- **Repairing**: I don't think there's any way for an instance to be in this state currently but in theory it's to indicate some sort of automated recovery attempt.
- **Failed**: The instance encountered an error at some point.
- **Destroyed**: The instance was deleted (but still exists in the DB, with `time_deleted` set).

The intention of introducing `Provisioned/ing` was to describe a step in-between `Creating/Stopped` and `Starting` (which automatically proceeds to `Running`). That is an instance exists on a sled, has its NICs and disks attached, and a corresponding propolis instance ready and waiting. (Propolis being ready so we can attach to the serial console but not yet running the vCPUs was the impetus for this change to start with).

Thinking more on this, I think we should stop leaking propolis' internal `Boot` state. From an external perspective, it's not much different than `Running`. In which case I think we can just rename `Provisioning` to `Starting`. I would like to keep this intermediate step between `Creating/Stopped` and `Running` because it can take a moment as everything starts up and it's nice to at least give the user some indication that it is in process.

I think the current hurdle here is how can we communicate a state after `Starting` but before `Running` (what I called `Provisioned` here) and how to allow such an action. It's somewhat hard because we currently use "Start" to indicate we want the instance to not only be setup but also running. So using `Started` I think would be confusing. Maybe `Setup`?

Started an RFD to try to clarify this in general: https://github.com/oxidecomputer/rfd/tree/0315/rfd/0315

Reply:
What about a `Suspended` state which means that the CPU and memory resources are allocated and consumed, but that the CPUs are not executing? I think we'll eventually need to describe the state machine in the user docs. I would suggest the following:

```
  +-----------------------------+
  v                             |
+-----------+     +----------+  |
| Suspended | <-- | Creating |  |
+-----------+     +----------+  |
  |                 |           |
  |                 |           |
  |                 v           |
  |               +----------+  |
  |               | Starting |  |
  |               +----------+  |
  |                 |           |
  |                 |           |
  |                 v           |
  |               +----------------------------+
  +-------------> |          Running           |
                  +----------------------------+
                    |           ^    |
                    |           |    |
                    v           |    v
                  +----------+  |  +-----------+
                  | Stopping |  +- | Rebooting |
                  +----------+     +-----------+
                    |
                    |
                    v
                  +----------+
                  | Stopped  |
                  +----------+
```

<details>
 <summary>dot</summary>
```
digraph {
    Creating -> Starting;
    Starting -> Running;
    Running -> Stopping;
    Running -> Rebooting;
    Rebooting -> Running;
    Stopping -> Stopped;
    Creating -> Suspended;
    Running -> Suspended;
    Suspended -> Running;  
}
```
</details>

This does not include the `Failed` or `Destroyed` states -- I thought it would just confuse the graph, but those are necessary states (of course!).

Note that this omits `Provisioning` and `Provisioned` as I don't think those are significant states to the end user and can be collapsed into `Starting`. I may be wrong, so please correct me.

I also omitted `Migrating` as another kind of `Running`. The end user literally should not need to know that this is happening. We should be transparent about that sub-state of `Running`, but otherwise I'd imagine a lot of code that needs to say `state == Running | Migrating`.

What would the `Repairing` state indicate?

Reply:
I like the state diagram, but should we move this discussion over to RFD 0315?

Reply:
Yup, spoke with adam earlier and plan to flesh out the RFD and have further discussion there. For now, this PR punts on the instance state stuff and is just for creating stopped instances.